ertviz/controllers/observation_response_controller.py

```diff
@@ -0,0 +1,177 @@
+import plotly.graph_objects as go
+from dash.dependencies import Input, Output, State
+from dash.exceptions import PreventUpdate
+from copy import deepcopy
+from ertviz.models import (
+    ResponsePlotModel,
+    BoxPlotModel,
+    PlotModel,
+    EnsembleModel,
+    MultiHistogramPlotModel,
+)
+import ertviz.assets as assets
+
+
+def _get_univariate_misfits_plots(misfits_df, color):
+    if misfits_df is None:
+        return []
+    style = deepcopy(assets.ERTSTYLE["observation-response-plot"]["misfits"])
+    style["line"]["color"] = color
+    style["marker"]["color"] = color
+    x_axis = misfits_df.pop("x_axis")
+    misfits_data = list()
+    for misfits in misfits_df:
+        plot = PlotModel(
+            x_axis=x_axis,
+            y_axis=misfits_df[misfits].pow(1 / 2.0).values,
+            text=misfits,
+            name=misfits,
+            **style,
+        )
+        misfits_data.append(plot)
+    return misfits_data
+
+
+def _get_univariate_misfits_boxplots(misfits_df, color):
+    if misfits_df is None:
+        return []
+    x_axis = misfits_df.pop("x_axis")
+    misfits_data = list()
+    for misfits in misfits_df.T:
+        plot = BoxPlotModel(
+            x_axis=[x_axis.loc[misfits]],
+            y_axis=misfits_df.T[misfits].values,
+            text=misfits,
+            name=f"Misfits@{int(x_axis.loc[misfits])}",
+            color=color,
+        )
+        misfits_data.append(plot)
+    return misfits_data
+
+
+def _get_observation_plots(observation_df, x_axis):
+    data = observation_df["values"]
+    stds = observation_df["std"]
+    x_axis = observation_df["x_axis"]
+    style = deepcopy(assets.ERTSTYLE["observation-response-plot"]["observation"])
+    observation_data = PlotModel(
+        x_axis=x_axis,
+        y_axis=stds.values,
+        text="Observations",
+        name="Observations",
+        **style,
+    )
+    return [observation_data]
+
+
+def _create_misfits_plot(response, selected_realizations, color):
+
+    x_axis = response.axis
+    # realizations = _get_univariate_misfits_plots(
+    realizations = _get_univariate_misfits_boxplots(
+        response.univariate_misfits_df(selected_realizations),
+        color=color,
+    )
+    ensemble_plot = ResponsePlotModel(
+        realizations,
+        [],
+        dict(
+            xaxis={"title": "Index"},
+            yaxis={"title": "Unit TODO"},
+            hovermode="closest",
+            uirevision=True,
+        ),
+    )
+    return ensemble_plot
+
+
+def observation_response_controller(parent, app):
+    @app.callback(
+        Output(parent.uuid("response-selector"), "options"),
+        [Input(parent.uuid("ensemble-selection-store"), "data")],
+    )
+    def _set_response_options(selected_ensembles):
+        # Should either return a union of all possible responses or the other thing which I cant think of...
```

Comment:
🤔

Reply:
Removed the comment for now as this represents only a list of selected ensembles